### ISSUE TITLE
Reflection mechanism for rewriting

### ISSUE THREAD

**Issue by Trebor-Huang** [agda/agda#6088]
Would it be reasonable to add reflection support for the `REWRITE` pragma? A few features that I can think of:
- Allow tactics to read the active rewriting equations (and also the relations declared as `BUITIN REWRITE`).
- Allow declaration of rewriting equations. This would be useful if someone is to prototype a feature in Agda, and needs systematic generation of reduction rules, e.g. HITs when there were no HITs, or H.O.T.T. that is currently under development, or other bizarre types I can't think of. 
- Allow enabling and disabling rewrites, for example only enabling some relevant rewrites in a subterm may greatly speed up typechecking. This could also allow tactics to decide which relations to use smartly. Even if this can't be done, disabling some equations during tactic execution would be useful for optimization.

**Comment by jespercockx** [comment]
Yes, I think this would definitely make sense. 1 and 2 are obviously useful, while 3 would require a bit of careful design to avoid problems with subject reduction. Perhaps it could make use of a similar mechanism to the already existing `dontReduceDefs` and `onlyReduceDefs` primitives?

**Comment by Trebor-Huang** [comment]
Would anything break if we allow a reduction rule to be enabled and disabled? `with` clauses sometimes don't work when the type is a reducible definition, so it might break?

**Comment by nad** [comment]
> Would anything break if we allow a reduction rule to be enabled and disabled?

Subject reduction?

**Comment by jespercockx** [comment]
Note that the `reduce` primitive of the reflection API is just a function producing a reflected `Term` that will be typechecked when it is unquoted, (with the full set of rewrite rules) so there is no risk of losing subject reduction here.
